### DISCUSSION TITLE
How can I use ESMFold with ESM2 dropout layers active at inference?

### DISCUSSION THREAD

**Discussion by FinnOD** [facebookresearch/esm#495] [Q&A]
Hi ESM Team, 

I want to produce multiple structures with the internal dropout layers active in order to build a distribution of structures (Drawing from [Gal and Ghahramani, 2015](https://arxiv.org/abs/1506.02142)). 

Currently I have just modified the example in the readme from `eval` to `train` like this:
```python
import torch
import esm

model = esm.pretrained.esmfold_v1()
model = model.train().cuda()
```

However I have some concerns that from inspecting the code this is only activating the dropout in ESMFold portion, not ESM. I believe the remaining dropout in ESM2 is in the  ESM2 -> TransformerLayer -> MultiheadAttention, but it is set to zero. 

I understand how I could change this by forking the repo, but how would I load the saved weights into my slightly adjusted model? 

Thanks for your help, 
Finn

**Comment by flix42** [comment]
Did you find a way to enable dropout?
I'm trying to get some variability from ESM2, but not calling eval or explicitly calling train dosnt seem to work.
Last line in the code snippet always returns true.
```
model, alphabet = esm.pretrained.esm2_t33_650M_UR50D()
[...]
# no model.eval()
model.train()
[...]
with torch.no_grad():
    results_dropout = [model(batch_tokens, repr_layers=[33], return_contacts=True) for i in range(3)]

(results_dropout[0]["representations"][33] == results_dropout[2]["representations"][33]).all()
```

**Comment by tomsercu** [comment]
The paper mentions:
>In ESM-2, we have made multiple small modifications to ESM-1b with the goal of increasing the effective capacity. ESM-1b had dropout both in hidden layers and attention which we removed completely to free up more capacity. In our experiments, we did not observe any significant perfor- mance regressions with this change.

An alternative idea to achieve what you are trying to achieve, is to randomly mask the sequence multiple times and make predictions on the different randomized inputs.
